src/sentry/ratelimits/cardinality.py

```diff
@@ -0,0 +1,135 @@
+import time
+from dataclasses import dataclass
+from typing import Collection, Optional, Sequence, Tuple
+
+from sentry.utils.services import Service
+
+Hash = str
+Timestamp = int
+
+
+@dataclass(frozen=True)
+class Quota:
+    # The number of seconds to apply the limit to.
+    window_seconds: int
+
+    # A number between 1 and `window_seconds`. Since `window_seconds` is a
+    # sliding window, configure what the granularity of that window is.
+    #
+    # If this is equal to `window_seconds`, the quota resets to 0 every
+    # `window_seconds`.  If this is a very small number, the window slides
+    # "more smoothly" at the expense of having much more redis keys.
+    #
+    # The number of redis keys required to enforce a quota is `window_seconds /
+    # granularity_seconds`.
+    granularity_seconds: int
+
+    #: How many units are allowed within the given window.
+    limit: int
+
+    def __post__init__(self) -> None:
+        assert self.window_seconds % self.granularity_seconds == 0
+
+
+@dataclass(frozen=True)
+class RequestedQuota:
+    # A string that all redis state is prefixed with. For example
+    # `sentry-string-indexer` where 123 is an organization id.
+    prefix: str
+
+    # A unit is an abstract term for the object type we want to limit the
+    # cardinality of.
+    #
+    # For example, if you want to limit the cardinality of timeseries in a
+    # metrics service, this would be a set of hashes composed from `(org_id,
+    # metric_name, tags)`.
+    #
+    # ...though you can probably omit org_id if it is already in the prefix.
+    unit_hashes: Collection[Hash]
+
+    # Which quotas to check against. The number of not-yet-seen hashes must
+    # "fit" into all quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    request: RequestedQuota
+
+    # The subset of hashes provided by the user `self.request` that were
+    # accepted by the limiter.
+    granted_unit_hashes: Collection[Hash]
+
+    # If len(granted_unit_hashes) < len(RequestedQuota.unit_hashes), this
+    # contains the quotas that were reached.
+    reached_quotas: Sequence[Quota]
+
+
+class CardinalityLimiter(Service):
+    """
+    A kind of limiter that limits set cardinality instead of a rate/count.
+
+    The high-level concepts are very similar to `sentry.ratelimits.sliding_windows`.
+
+    Instead of passing in numbers and getting back smaller numbers, however, the
+    user passes in a set and gets back a smaller set. Set elements that have
```

Comment:
Food for thoughts. Not necessarily for this PR.
I think we could have a common interface between the different types of rate limiters. 
- Both rate limiter classes are sliding window and rely on a GrantedQuota and RequestedQuota.
- The only difference is the data type that represent the quota itself. One rate limiter request ints and receives smaller ints, the other requests sets and receives smaller sets. I wonder if the interface can be generic with respect to the quota type.

Reply:
yes, i erred on the side of code duplication for now because i am not really confident that the interface is stable when we get to implementing backends. if the interface stays stable after productionizing we should consider it